propagation/http_trace_context_propagator.go

```diff
@@ -0,0 +1,128 @@
+// Copyright 2019, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package propagation
+
+import (
+	"context"
+	"encoding/hex"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"go.opentelemetry.io/api/trace"
+
+	"go.opentelemetry.io/api/core"
+	apipropagation "go.opentelemetry.io/api/propagation"
+)
+
+const (
+	supportedVersion  = 0
+	maxVersion        = 254
+	traceparentHeader = "traceparent"
+)
+
+type httpTraceContextPropagator struct{}
+
+var _ apipropagation.TextFormatPropagator = httpTraceContextPropagator{}
+
+func (hp httpTraceContextPropagator) Inject(ctx context.Context, supplier apipropagation.Supplier) {
+	sc := trace.CurrentSpan(ctx).SpanContext()
+	if sc.IsValid() {
+		h := fmt.Sprintf("%.2x-%.16x%.16x-%.16x-%.2x",
+			supportedVersion,
+			sc.TraceID.High,
+			sc.TraceID.Low,
+			sc.SpanID,
+			sc.TraceOptions)
+		supplier.Set(traceparentHeader, h)
+	}
+}
+
+func (hp httpTraceContextPropagator) Extract(ctx context.Context, supplier apipropagation.Supplier) core.SpanContext {
+	h := supplier.Get(traceparentHeader)
+	if h == "" {
+		return core.EmptySpanContext()
+	}
+
+	sections := strings.Split(h, "-")
+	if len(sections) < 4 {
+		return core.EmptySpanContext()
+	}
+
```

Comment:
Should we really check if all the characters in at least first four sections are either digits or lowercase letters from a to f? That's what the [W3C trace context spec](https://www.w3.org/TR/trace-context/) says in several places, like

> `HEXDIGLC = DIGIT / "a" / "b" / "c" / "d" / "e" / "f" ; lowercase hex character`

or

> Vendors MUST ignore the `traceparent` when the `parent-id` is invalid (for example, if it contains non-lowercase hex characters).